text/1010-global-vars.rst

```diff
@@ -0,0 +1,179 @@
+
+
+Specification
+=============
+
+A global variable must be declared before it can be used in any expression.
+
+CREATE GLOBAL
+-------------
+
+Define a new global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    CREATE [ REQUIRED ] [ MULTI | SINGLE] GLOBAL <name> -> <type> [ "{"
+        <subcommand> ; [...]
+    "}" ] ;
+
+    # Computed global variable form:
+
+    [ WITH <with-item> [, ...] ]
+    CREATE [ REQUIRED ] GLOBAL <name> := <expr> ;
+
+    # where <subcommand> is one of
+
+      SET default := <value>
+      RESET default
+      RENAME TO <newname>
+      USING <expr>
+      RESET EXPRESSION
+      CREATE ANNOTATION <annotation-name> := <value>
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+The ``<type>`` of non-computed globals must be a primitive type, i.e. it
+*must not* be an object type or a collection of object types.  This is because
+object type variables will require some form of referential integrity checks
+like links do (i.e. the ``on target delete`` spec), which complicates the
+implementation considerably.  Computed globals can have any type, and so it
+is possible to define an object type global in terms of its ``id``, e.g::
+
+    CREATE GLOBAL user_id -> uuid;
+    CREATE GLOBAL user := (SELECT User FILTER .id = (GLOBAL user_id));
+
+If a global is declared as ``REQUIRED``, it *must* provide a default value, or,
+if the global is computed, the expression must be non-optional.
+
+Explicit cardinality is specified via the ``SINGLE`` or ``MULTI`` keyword,
+otherwise the cardinality is assumed to be ``SINGLE`` in non-computed globals,
+and is inferred from the expression in computed globals.
+
+
+ALTER GLOBAL
+------------
+
+Alter the definition of a global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    ALTER GLOBAL <name>
+    "{" <subcommand>; [...] "}" ;
+
+    # where <subcommand> is one of
+
+      SET default := <value>
+      RESET default
+      RENAME TO <newname>
+      USING <expr>
+      RESET EXPRESSION
+      SET REQUIRED
+      DROP REQUIRED
+      SET SINGLE
+      SET MULTI
+      CREATE ANNOTATION <annotation-name> := <value>
+      ALTER ANNOTATION <annotation-name> := <value>
+      DROP ANNOTATION <annotation-name>
+
+
+DROP GLOBAL
+-----------
+
+Remove a global variable.
+
+Synopsis::
+
+    [ WITH <with-item> [, ...] ]
+    DROP GLOBAL <name> ;
+
+
+SET GLOBAL
+----------
+
+Set the value of a non-computed global variable *in the current session*
+by evaluating the given expression.
+
+Synopsis::
+
+    SET GLOBAL <name> := <expr> ;
+
+
+RESET GLOBAL
+------------
+
+Reset a non-computed global variable to its default value.
+
+Synopsis::
+
+    RESET GLOBAL <name> ;
+
+
+Referring to globals in queries
+===============================
+
+The new ``GLOBAL <name>`` expression is used to refer to the value of the
+given global variable in queries.  For example::
+
+    SELECT User FILTER .id = (GLOBAL user_id)
```

Comment:
What's the precedence? Always using lisp-style parenthesis is not convenient.

Alternative might be something like `$global.user_id`.

Reply:
The proposed precedence is like a unary minus, so the parentheses are optional.  

`$global.user_id` looks like an object argument `$global` with a `user_id` property.  It's OK to think of all globals as properties/links of a `Global` singleton, but we shouldn't indicate it using the argument syntax, I think.  We can use the same approach as for `Config`, but then we would need to add a mechanism to indicate singleton objects in schema, so that cardinality inference of `Global.user_id` is correct.